oio/common/green.py

```diff
@@ -77,6 +78,158 @@ def ratelimit(run_time, max_rate, increment=1, rate_buffer=5):
+                break
+            curr_partition = partition
+    else:
+        curr_partition = policy[0]
+
+    return curr_partition[1]
+
+
+def ratelimit_function_next_rate(curr_date, policy):
+    """
+    Given a current date and a policy, calculate the date at which the next
+    rate change is scheduled.
+
+    (Could be useful if the rate limited operation is fast, and as such we
+    would want to cache the next rate date so that instead of selecting the
+    rate each op, we'd just compare to a timestamp and return a cached value,
+    which in the current implementation would make it go from a for loop with
+    several comparisons to about a single comparison)
+
+    :param curr_date: The current datetime
+    :type curr_date: `datetime`
+    :param policy: A list representing a validated policy.
+    :returns: the next scheduled rate and the `datetime` object for the next
+              scheduled rate change.
+    """
+    next_day = False
+    for partition in policy:
+        curr_partition = partition
+        if (curr_date - partition[0]).date() < curr_date.date():
+            break
+    else:
+        curr_partition = policy[0]
+        next_day = True
+    next_date = datetime(curr_date.year, curr_date.month, curr_date.day)
+    next_date += curr_partition[0]
+    if next_day:
+        next_date += timedelta(days=1)
+    return curr_partition[1], next_date
+
+
+def ratelimit_policy_from_string(policy_str):
+    """
+    :rtype: `list` of 2-tuples with a `datetime.timedelta` and an integer.
+    """
+    policy = list()
+    changes = policy_str.split(';')
+    for change in changes:
+        try:
+            time_str, rate_str = change.split(':', 1)
+            hour_str, min_str = time_str.split('h', 1)
+            td = timedelta(hours=int(hour_str), minutes=int(min_str))
+            rate = int(rate_str)
+        except ValueError as err:
+            raise ValueError("Unparseable rate change '%s': %s" %
+                             (change, err))
+        policy.append((td, rate))
+    policy.sort()
+    return policy
+
+
+def ratelimit_function_build(policy):
+    """
+    Given a policy, return a customized wrapper around ratelimit for a
+    time aware rate limiter.
+    :param policy: An array representing a rate limiting policy as described
+                    by ratelimit_validate_policy.
+    :return: A callable function similar in signature to ratelimit but that
+             ignores all parameters other than the first one.
+    """
+    if isinstance(policy, basestring):
+        policy = ratelimit_policy_from_string(policy)
+    ratelimit_validate_policy(policy)
+
+    def _ratelimiter(run_time, *args, **kwargs):
+        """
+        The ratelimit wrapper that takes into account the custom policy, and
+        ignores all the other parameters other than run_time
+        :param run_time: The last time the operation was executed in seconds.
+        """
+        curr_date = datetime.today()
```

Comment:
To not call twice in a row `time.time()`, we could store the value and transmit it to `ratelimit`.
```suggestion
        curr_time = time.time()
        curr_date = datetime.fromtimestamp(curr_time)
```

Reply:
Thanks for the suggestion.